### PR TITLE
add functions to decode an epoch timestamp

### PR DESCRIPTION
**Pull request by marler8997**
The code added here is alternative to the libc `gmtime` function.  This function takes a unix epoch timestamp and decodes it into things like the year/month/day/time/etc.  I looked at various libc implementations to see how it was implemented and this appears to be correct (the tests also match what I get on epochconverter.com).

I split the single `gmtime` function into multiple functions based on the order in which the date information is decoded so that applications can limit their calculations to what they need rather than calculating everything in a single function.  The data structures are based on the order that pieces of the timestamp are calculated.

### CODE REVIEW

File: lib/std/time/epoch.zig [hunk 1]
```diff
@@ -40,3 +43,179 @@ pub const morphos = amiga;
...
+        .oct => 31,
+        .nov => 30,
+        .dec => 31,
+    };
+}
+
+pub const YearAndDay = struct {
+    year: Year,
+    /// The number of days into the year (0 to 365)
+    day: u9,
```
Comment (Vexu) [inline lib/std/time/epoch.zig:119]: ISO 8601 says that day of the year is in range `[1,365 (366)]`, maybe this should be too?

File: lib/std/time/epoch.zig [hunk 2]
```diff
@@ -40,3 +43,179 @@ pub const morphos = amiga;
...
+}
+
+pub fn getDaysInYear(year: Year) u9 {
+    return if (isLeapYear(year)) 366 else 365;
+}
+
+pub const YearLeapKind = enum(u1) { not_leap, leap };
+
+pub const Month = enum(u4) {
+    jan,
```
Comment (LemonBoy) [inline lib/std/time/epoch.zig:75]:
```suggestion
    jan = 1,
```

And so forth. You can avoid the +1 dance and the `numeric` helper.

**Comment by LemonBoy** [comment]
This is possibly related to #8396 as it implements random chunks of a `Date` structure.

File: lib/std/time/epoch.zig (hunk 2 above)
Comment (marler8997) [reply lib/std/time/epoch.zig]: sure that should work, done

File: lib/std/time/epoch.zig (hunk 1 above)
Comment (marler8997) [reply lib/std/time/epoch.zig:119]: Not sure.  I actually can't think of a use case for wanting the day of the year. For the calculation we need it to be 0 based, so if we make it 1 based, we'll just be adding 1 then undoing that operation in `calculateMonthDay`. An alternative is also to create a method that will return this day value with 1 added. Not sure what the best route is here.

**Review by Vexu — APPROVED** [review]

**Pull request merged**